i2c: Add RTIO-compatible driver for nRFx TWI

ubieda opened the pull request:
### Description
- Add first version of nrfx_twi_rtio.c, the RTIO-compatible driver for nrfx devices.
- Add overlay to make the i2c_ram test compatible with nRF52840DK, the board used to develop and test against.

### Testing
- Build command: `west build -b nrf52840dk/nrf52840 tests/drivers/i2c/i2c_ram -- -DCONFIG_I2C_RTIO=y`
- Output:
``` console
*** Booting Zephyr OS build v3.6.0-3653-gf246797ccd33 ***
Running TESTSUITE i2c_ram
===================================================================
START - test_ram_rtio
submitting write from thread 0x20000250 addr 7
 PASS - test_ram_rtio in 0.007 seconds
===================================================================
START - test_ram_rtio_isr
timer submitting write, addr e
timer checking write result, submitting read
read command complete
read data complete
 PASS - test_ram_rtio_isr in 0.015 seconds
===================================================================
START - test_ram_transfer
ram using i2c_transfer from thread 0x20000250 addr 15
 PASS - test_ram_transfer in 0.008 seconds
===================================================================
START - test_ram_write_read
ram using i2c_write and i2c_write_read from thread 0x20000250 addr 1c
 PASS - test_ram_write_read in 0.009 seconds
===================================================================
TESTSUITE i2c_ram succeeded

------ TESTSUITE SUMMARY START ------

SUITE PASS - 100.00% [i2c_ram]: pass = 4, fail = 0, skip = 0, total = 4 duration = 0.039 seconds
 - PASS - [i2c_ram.test_ram_rtio] duration = 0.007 seconds
 - PASS - [i2c_ram.test_ram_rtio_isr] duration = 0.015 seconds
 - PASS - [i2c_ram.test_ram_transfer] duration = 0.008 seconds
 - PASS - [i2c_ram.test_ram_write_read] duration = 0.009 seconds

------ TESTSUITE SUMMARY END ------

===================================================================
PROJECT EXECUTION SUCCESSFUL
```

ubieda committed:
i2c: Add nRFX TWI RTIO-compatible driver

First version of RTIO-compatible nrfx_twi driver.

Test Setup:
- Board: nrf52840dk
- Test: `tests/drivers/i2c/i2c_ram`
- Additional Kconfig: `CONFIG_I2C_RTIO=y`

Signed-off-by: Luis Ubieda <luisf@croxel.com>

ubieda committed:
tests: i2c: i2c_ram: Add nrf52840dk overlay

To make it compatible with this test.

Signed-off-by: Luis Ubieda <luisf@croxel.com>

ubieda commented:
CC @carlescufi

teburd commented on drivers/i2c/i2c_nrfx_twi_rtio.c:
```diff
@@ -0,0 +1,365 @@
+	struct i2c_rtio *ctx;
+	struct k_sem transfer_sync;
+	struct k_sem completion_sync;
+	volatile nrfx_err_t res;
```
Is this still needed or is the error returnable more directly with rtio_iodev_sqe_err()?

ubieda replied:
No longer needed. Removed!

teburd commented on drivers/i2c/i2c_nrfx_twi_rtio.c:
```diff
@@ -0,0 +1,365 @@
+	int ret = 0;
+
+	/** Enabling while already enabled ends up in a failed assertion: skip it. */
+	if (k_sem_take(&dev_data->completion_sync, K_NO_WAIT) == 0) {
```
start should only be called if nothing else is in progress, is this semaphore really needed?

ubieda replied:
The logic followed in this code attempts to replicate this sequence: https://github.com/zephyrproject-rtos/zephyr/blob/2c6306d09954de13f993ebc28e173246b9b6ef39/drivers/i2c/i2c_nrfx_twi.c#L55

consisted of:
1. Enable the TWI periheral before starting the set of messages.
2. Go through each message, transferring one after the other.
3. Disable the TWI peripheral.

In this driver, `i2c_nrfx_twi_msg_start()` is called on every message, and the semaphore signals whether it's the beginning of the set of messages.

LMK if you think we should structure it differently.

teburd replied:
Ah, you could use a simple boolean flag I believe in this case as a data member, semaphore are lightweight but a boolean seems sufficient in this case, no threads are going to be waiting on this semaphore

ubieda replied:
You're right - I've changed this to a boolean named `twi_enabled`.

teburd commented on drivers/i2c/i2c_nrfx_twi_rtio.c:
```diff
@@ -0,0 +1,365 @@
+	}
+
+	if (!ret) {
+		res = nrfx_twi_xfer(&config->twi, &cur_xfer, xfer_flags);
```
Is there no interrupt to catch on completion of a transfer?

ubieda replied:
`res` here just signals if the request started successfully. `event_handler()` triggers when the transfer completes (either successfully or in an error).

Does that make sense?

teburd commented on drivers/i2c/i2c_nrfx_twi_rtio.c:
```diff
@@ -0,0 +1,365 @@
+		(void)i2c_nrfx_twi_start(dev);
+	} else {
+		nrfx_twi_disable(&config->twi);
+		k_sem_give(&data->completion_sync);
```
i don't think this semaphore is needed anymore, no threads are waiting on it

ubieda replied:
Resolution depends on https://github.com/zephyrproject-rtos/zephyr/pull/72539#discussion_r1596846675

teburd commented on drivers/i2c/i2c_nrfx_twi_rtio.c:
```diff
@@ -0,0 +1,365 @@
+	case RTIO_OP_I2C_CONFIGURE:
+		(void)i2c_nrfx_twi_configure(dev, sqe->i2c_config);
+		return false;
+	default:
```
RECOVER should also be in here, there's an op for it and a helper that can be called by the i2c API implementer https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/i2c/i2c_rtio.c#L193

ubieda replied:
Agreed - Added support for `RTIO_OP_I2C_RECOVER` now.

teburd reviewed (commented):
Looks really good, and I love that it was ~300 lines of code change to make it happen. Some minor comments related to I believe no longer needed semaphores and the matching recover API implementation. +1 on correcting those from me, thank you for contributing this!

ubieda commented on drivers/i2c/i2c_nrfx_twi_rtio.c:
```diff
@@ -0,0 +1,365 @@
+
+struct i2c_nrfx_twi_data {
+	struct i2c_rtio *ctx;
+	struct k_sem transfer_sync;
```
`transfer_sync` no longer needed. Removed!

ubieda commented:
Pushed iteration of this PR with the following changes:
- Drop support for `res` and `transfer_sync` from `struct i2c_nrfx_twi_data` as they're no longer needed.
- Add support for `RTIO_OP_I2C_RECOVER`.

carlescufi commented:
@anangl @masz-nordic @nika-nordic @kl-cruz @magp-nordic please take a look

masz-nordic commented on drivers/i2c/i2c_nrfx_twi_rtio.c:
```diff
@@ -0,0 +1,362 @@
+#define I2C_TRANSFER_TIMEOUT_MSEC K_FOREVER
+#endif
+
+struct i2c_nrfx_twi_data {
```
```suggestion
struct i2c_nrfx_twi_rtio_data {
```
Same in other instances where naming clashes with one from `i2c_nrfx_twi.c`.

ubieda replied:
Request addressed.

masz-nordic commented on drivers/i2c/i2c_nrfx_twi_rtio.c:
```diff
@@ -0,0 +1,362 @@
+	uint32_t dev_config;
+};
+
+struct i2c_nrfx_twi_config {
+	nrfx_twi_t twi;
+	nrfx_twi_config_t config;
+	const struct pinctrl_dev_config *pcfg;
+};
```
It would be great to have code shared between this driver and `i2c_nrfx_twi.c` extracted to a separate file.
From what I see that would be config, most of `i2c_nrfx_twi_msg_start`, `i2c_nrfx_twi_configure`, `i2c_nrfx_twi_recover_bus`, `init_twi`, `twi_nrfx_pm_action` and most of device defines.

ubieda replied:
Request addressed. The following APIs have been refactored:
- `struct i2c_nrfx_twi_config`
- `init_twi` -> `i2c_nrfx_twi_init`
- `i2c_nrfx_twi_configure`
- `i2c_nrfx_twi_recover_bus`
- `twi_nrfx_pm_action`
- Body of `i2c_nrfx_twi_msg_start` -> `i2c_nrfx_twi_msg_transfer`
- Device Macros.

ubieda committed:
i2c: nrfx_twi: Refactor nrfx_twi to utilize common code

Code is refactored to take advantage of common codebase shared with
i2c_nrfx_twi_rtio driver.

Signed-off-by: Luis Ubieda <luisf@croxel.com>

ubieda commented:
Update:
- Renamed RTIO dev data structure to `struct i2c_nrfx_twi_rtio_data`
- Refactored common code between `i2c_nrfx_twi_rtio.c` and `i2c_nrfx_twi.c`, now used on both drivers.

masz-nordic commented on drivers/i2c/i2c_nrfx_twi_common.c:
```diff
@@ -0,0 +1,152 @@
+	if (flags & I2C_MSG_ADDR_10_BITS) {
+		LOG_ERR("10-bit I2C Addr devices not supported");
+		ret = -ENOTSUP;
+	} else if (!(flags & I2C_MSG_STOP)) {
```
This part is missing:
```
			/* - if the transfer consists of more messages
			 *   and the I2C repeated START is not requested
			 *   to appear before the next message, suspend
			 *   the transfer after the current message,
			 *   so that it can be resumed with the next one,
			 *   resulting in the two messages merged into
			 *   a continuous transfer on the bus
			 */
			if ((i < (num_msgs - 1)) &&
			    !(msgs[i + 1].flags & I2C_MSG_RESTART)) {
				xfer_flags |= NRFX_TWI_FLAG_SUSPEND;
                        }
```

ubieda replied:
That's true. Thanks for pointing it out. I was struggling to refactor this into common code.

I have now resolved to add a flag `bool more_msgs` to represent this condition in `i2c_nrfx_twi_msg_transfer`.

ubieda commented:
Update:
- Add `bool more_msgs` to cover this in common code:
``` c
			if ((i < (num_msgs - 1)) &&
			    !(msgs[i + 1].flags & I2C_MSG_RESTART)) {
				xfer_flags |= NRFX_TWI_FLAG_SUSPEND;
                        }
```

ubieda commented on drivers/i2c/i2c_nrfx_twi_common.c:
```diff
@@ -0,0 +1,173 @@
+			ret = -ENOTSUP;
+		} else {
+			xfer_flags |= NRFX_TWI_FLAG_TX_NO_STOP;
+		}
```
Apologies - I failed to remove this on the latest iteration. It's now addressed.

masz-nordic reviewed (approved):
A few last suggestions to establish proper namespacing between those two drivers.
Not a must have though.

ubieda commented:
@masz-nordic Thanks for your feedback. I'll follow-up on a separate PR to apply these suggestions in pro of establishing namespacing between the two drivers.
